keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,444 @@
+loop. The reconciliations are batched by 1s periods to make sure multiple
+created pods at the similar time don't trigger too many reconciliations.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes
+necessary to implement this enhancement.
+
+##### Prerequisite testing updates
+
+#### Unit Tests
+
+We add integration tests for the following scenarios:
+- a new node is added making workload admission possible
+- the PodSet assignment is computed successfully for `require-topology`
+- the PodSet assignment cannot be computed for `require-topology`
+- the PodSet assignment is computed successfully for `prefer-topology` across
+ multiple values
+- the PodSet assignment cannot be computed for `prefer-topology`
+- the schedulingGate is added to the pod template
+
+#### Integration tests
+
+We are going to add the integration tests to make sure the implementation is
+well covered. In particular, the following scenarios need coverage:
+- adding new nodes
+- PodSet assignment for `prefer-topology` and `require-topology`
+- Job's pod template has the "topology" scheduling gate injected
+- un-gate pods by with the "topology" scheduling gate
+- the Workload can be suspended and unsuspended
+
+### Graduation Criteria
+
+#### Alpha (MVP):
+
+- support for all built-in integrations: Job and JobSet
+- support single-level hierarchy
+- support TAS with minimal cross to other features (no cohorts, no preemption,
+  no reclaimable pods)
+
+The new validations which are for MVP, but likely will be relaxed in the future:
+- ClusterQueue is marked inactive if it contains a TAS ResourceFlavor and
+  belongs to a cohort
+- ClusterQueue is marked inactive if it contains a TAS ResourceFlavor and
+  enables preemptions
+- ClusterQueue is marked inactive if it contains a TAS ResourceFlavor and uses
+  MultiKueue admission check
+- ClusterQueue is marked inactive if it contains a TAS ResourceFlavor and uses
+  ProvisioningRequest admission check
+
+#### Beta
+
+- support for all other integrations
+- support multi-level hierarchy
+- support replacement timeout in WaitForPodsReady
+
+#### Stable
+
+Consider the following improvements and implement if feasible:
+- put pods with consecutive indexes of an IndexedJob on close nodes
+- support the following features: reclaimable, pods, cohorts preemption
+ within cluster queue
+- support re-computing the TopologyAssignment while the workload
+ is running
+- perform full scheduling simulation rather than just capacity counting
+ (including pod affinities and anti-affinities)
+
+## Implementation History
+
+- 2024-07-25 - PR merged [Allow mutating schedulingGates when the Jobset is suspended](Allow mutating schedulingGates when the Jobset is suspended)
+- 2024-07-30 - this KEP
+
+
+## Drawbacks
+
+Tracking of nodes by Kueue will increase its code complexity and memory
+consumption.
+
+## Alternatives
```

Comment:
Could you add some comments about how this would inface with scheduler or kubelet based polciies?

https://scheduler-plugins.sigs.k8s.io/docs/kep/119-node-resource-topology-aware-scheduling/readme/

At the end of the day, users can specify this at the kueue level but if the scheduler or kubelet are not set up, I don't see how you can guarantee this at the Kueue level.

Reply:
Maybe there is an unnecessary naming coincidence, but I don't think the feature in Kueue is related to the plugin.
We don't require it to run. Also, this feature is not related to NUMA nodes. Let me know if I'm missing something.

The aim of this feature is to provide compact placement of pods on nodes, by using the information about node placement in datacenter.

Reply:
So this feature is not really about topology on a node (where Topology Manager or Topology-aware-schediling) would take in account. You are proposing intra node topology aware scheduling (node - node placement).

So from a Kueue follower hat, I think this is a neat feature.

Wearing from openshift hat, I am concerned that this feature is supporting a particular cloud vendor and not really publishing how one can label these nodes so that this feature could be used.

Are you planning on proposing official labels/annotations in k8s to walk through how other cloud provider vendors (or on-prem admins) can actually set up their cluster so that this feature is valid.

Reply:
And don't take this to be blocking but I would like to see some work being done to maybe formalize how a cluster should be labeled. Otherwise I don't know how other vendors/on-prem could use this feature.

Reply:
I attended a wg-serving meeting earlier today where someone was talking about the types of ways data centers are laid out and what is necessary for higher speed interconnects. And I think the point of that meeting was to start to define what kinds of labels/information would be necessary.

I think this is a hard problem..

Reply:
Networking in data centers is complex for sure.  Agree, this may proposal does not reflect every nuance of a data center networking. Especially, since details of networking, the depth of hierarchy will differ between data centers of different cloud providers. The details may even differ between multiple data centers of the same cloud provider. The details, I believe, may also differ between machine types: TPU vs GPU.

This is the first iteration of support for Node-node Topology Aware Scheduling in Kueue. Every journey needs to start with a first step.

Thus, the proposed model is deliberately simplified so that it could work as a common denominator between different cloud providers. 

Note that, the label names `cloud-provider.com/topology-block` and `cloud-provider.com/topology-rack` used in [Hierarchy organization](https://github.com/kubernetes-sigs/kueue/blob/5d7847bed87ffa353732164de229b0f94aeab8bd/keps/2724-topology-aware-schedling/README.md#hierarchy-representation) are just examples, and they can be customized in the [admin-facing API](https://github.com/kubernetes-sigs/kueue/blob/5d7847bed87ffa353732164de229b0f94aeab8bd/keps/2724-topology-aware-schedling/README.md#admin-facing-api) for a ResourceFlavor. 

The only thing that matters for the model is that there is some hierarchy of nodes in a data center, which is almost certainly true. Since almost every data center has some concept of racks and blocks I use such labels to call the intuition, but it is flexible. The mechanics of exposing information as labels are left to cloud providers.